SessionMainWindow Base class for saving and loading sessions

lauramurgatroyd opened the pull request:
This adds:
- `SessionsMainWindow.py` - which is a base class for our apps which create a session folder where any files generated in the app are saved, and provides the ability to permanently save and reload sessions.
- `SessionsMainWindow_example.py` - an example of using the SessionsMainWindow - you can run this example, change the state of widgets in the form, save the session, reload the session and see the state of the widgets be restored.
- `SessionsDialogs.py` - the dialogs used by the `SessionsMainWindow.py` 
- `io.py` - contains method for zipping a directory, used by `SessionsMainWindow.py`

paskino reviewed (changes requested):
Broadly speaking it looks great, 2 comments:

1. can we help the user with `addToMenu`
2. why do we need all those `os.chdir`?

lauramurgatroyd commented on eqt/ui/SessionMainWindow.py:
```diff
@@ -0,0 +1,675 @@
+
+        Notes
+        -----
+        You may wish to override this method to add additional menu options
```
In option 3, would createMenu still do the same thing as it currently does i.e. create the basic menu?

lauramurgatroyd replied:
Perhaps inside createMenu we should set self.menu_bar, self.menus, so that they can be retrieved in addToMenu?

paskino replied:
By name `createMenu` doesn't return anything, so I agree that `self.menu_bar` and `self.menu` should be set inside it.

In option 3, `createMenu` will do the same thing and then set `self.menu_bar` and `self.menu`

lauramurgatroyd replied:
I have now updated, as discussed

lauramurgatroyd commented on test/test_SessionDialogs.py:
```diff
@@ -0,0 +1,135 @@
+        sdsd.browse_for_dir.assert_called_once()
+
+    def test_browse_dialog_updates_session_directory_label(self):
+        example_dir = "C:\\Users\\test_user\\Documents\\test_dir"
```
Yes, we are not creating the directory, we are just testing if the name appears in the label

lauramurgatroyd commented on eqt/ui/SessionMainWindow.py:
```diff
@@ -0,0 +1,675 @@
+            self.createSessionsDirectorySelectionDialog(new_session=True)
+        else:
+            session_folder_name  = self.settings.value('sessions_folder')
+            os.chdir(session_folder_name)
```
No, as you found below.
In general, re: the chdirs:
In my apps I like to have the current working directory be the current session folder / current run folder when we are in a run. So the chdirs are making that happen.
However, I will move this out of the base class to simplify things

lauramurgatroyd commented on eqt/ui/SessionMainWindow.py:
```diff
@@ -0,0 +1,675 @@
+
+        if settings_name is None:
+            settings_name = title
+        self.settings = QSettings(settings_name, settings_name)
```
done

lauramurgatroyd commented on eqt/ui/SessionDialogs.py:
```diff
@@ -0,0 +1,175 @@
+
+        browse_button = QPushButton('Browse')
+        browse_button.clicked.connect(self.browse_for_dir)
+        self.addSpanningWidget(browse_button, 'browse_button')
```
I removed the spanning button

paskino commented on examples/SessionMainWindow_example.py:
```diff
@@ -26,6 +26,12 @@ def __init__(self, *args, **kwargs):
 
+    def addToMenu(self):
+        ''' Adds an example extra menu.'''
+        extra_menu = self.menu_bar.addMenu("Extra Menu")
+        self.menu_bar.addMenu(extra_menu)
```
I was very confused by these lines, however I see that the confusion comes from Qt
https://doc.qt.io/qt-6/qmenubar.html#public-functions
`addMenu` called with a string returns a `QMenu`, if called with a `QMenu` returns a `QAction`

paskino commented on eqt/ui/SessionDialogs.py:
```diff
@@ -0,0 +1,172 @@
+        self.Cancel.setText('New Session')
+
+class WarningDialog(QMessageBox):
+    def __init__(self, parent=None, message=None, window_title=None, detailed_text=None):
```
Can we change the signature to 
```
def __init__(self, parent=None, window_title=None, message=None, detailed_text=None)
```

It would be great if we can simply plug this in a `Worker` error signal as

```
my_worker.signals.error.connect(some_function_that_starts_the_warning_dialog, main_window, 'title of the warning')
```
then the tuple emitted by the error signal will populate the warning dialog.

lauramurgatroyd replied:
updated signature of WarningDialog.
Added example of adding an error connection to a worker.

lauramurgatroyd commented:
I am satisfied this works fine on linux, so ready to merge except for the small requested change

lauramurgatroyd commented:
Just updated the order of parameters for the WarningDialog and the ErrorDialog, modified createMenu so it doesn't return the attributes it is setting, autoformatted, and fixed some unit tests.
